Creation of remote audio rendering internal unit is needlessly complex

kkinnunen-apple opened the pull request:
#### 6dc0b32040a92f08d8a14b620c19fbd629d6ea12
<pre>
Creation of remote audio rendering internal unit is needlessly complex
<a href="https://bugs.webkit.org/show_bug.cgi?id=248168">https://bugs.webkit.org/show_bug.cgi?id=248168</a>
rdar://problem/102577160

Reviewed by Darin Adler.

It takes two Functions that capture the same creator instance.
The creation is configured with a third Function.

Replace the two Functions with just a normal Client interface.

Replace the global factory Function with a normal function pointer.

Rename the creating functions to be just normal Class::create(...).

* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererInternalUnit::setCreateFunction):
(WebCore::AudioMediaStreamTrackRendererInternalUnit::create):
(WebCore::AudioMediaStreamTrackRendererInternalUnit::createLocalInternalUnit): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::LocalAudioMediaStreamTrackRendererInternalUnit): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::retrieveFormatDescription): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::setAudioOutputDevice): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::start): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::stop): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::createAudioUnitIfNeeded): Deleted.
(WebCore::clipAudioBuffer): Deleted.
(WebCore::clipAudioBufferList): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::render): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::renderingCallback): Deleted.
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererUnit::AudioMediaStreamTrackRendererUnit):
(WebCore::AudioMediaStreamTrackRendererUnit::render):
(WebCore::getCreateInternalUnitFunction): Deleted.
(WebCore::AudioMediaStreamTrackRendererUnit::setCreateInternalUnitFunction): Deleted.
(WebCore::createInternalUnit): Deleted.
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.h:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::Unit):
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::reset):
(WebKit::renderCallback): Deleted.
(WebKit::resetCallback): Deleted.
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::notifyReset): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::createRemoteAudioMediaStreamTrackRendererInternalUnitProxy):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::Proxy):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::startThread):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::reset):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::createRemoteInternalUnit): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.h:
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::setupCaptureProcesses):

Canonical link: <a href="https://commits.webkit.org/256969@main">https://commits.webkit.org/256969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02791ca566e1245d14fcad8d294cde41c5accd3d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106906 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167172 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6963 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35394 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89795 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103582 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103050 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84019 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32229 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87090 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75154 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/664 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/647 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21831 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4804 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5453 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41180 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
```diff
@@ -48,12 +48,17 @@
 
 namespace WebCore {
 
+namespace {
```
In practice wrapping this in an anonymous namespace just makes identifiers trickier when debugging and does not have a practical benefit since no name conflict will occur. I suggest leaving this as is.

kkinnunen-apple replied:
But how is it, in practice, making debugging harder? IIRC every time I've set a breakpoint, it has worked for me by just using equivalent of `b LocalAudioMediaStreamTrackRendererInternalUnit`.

All compilation unit local functions, regardless of name conflict probability or not are consistently marked as static and reviews will consistently nag about this. Based on this, I would expect consistency for any function name, regardless if it is prefixed with a class name. If it is hard to debug the types, it should be hard to debug the free functions, too.

Marking the types as internal linkage has the benefit of :
 - Theoretically stops them from being considered in some template application cases (for which I cannot remember the specifics)
 - Theoretically makes the compilation a bit simpler since the names are internal
 - Theoretically enables more inlining as the compiler knows earlier that all use-sites are in the compilation unit only

I cannot currently prove any of these, so we might as well go with the original version without the anonymous namespace. It would be great if the reviews would be more predictable, though..

darinadler replied:
Yes, in WebKit we use `static` consistently for functions local to a source file; the unused function warning we get is one of the benefits. We typically do not use anonymous namespaces for functions or for classes local to a source file. We could start doing it but I agree that we’d have to confirm the benefits. If I am wrong about the debugging problem, then I am sorry for misleading you.

darinadler commented on Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
```diff
@@ -48,12 +48,17 @@
+    }
 
 private:
+    LocalAudioMediaStreamTrackRendererInternalUnit(Client&);
```
I suggest `explicit`.

darinadler commented on Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
```diff
@@ -290,6 +289,21 @@ OSStatus LocalAudioMediaStreamTrackRendererInternalUnit::renderingCallback(void*
 
+}
+
+static AudioMediaStreamTrackRendererInternalUnit::CreateFunction createInternalUnit = LocalAudioMediaStreamTrackRendererInternalUnit::create;
```
Does it work if you use `auto` here instead of writing out the type?

darinadler commented on Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
```diff
@@ -48,7 +48,7 @@ namespace WebKit {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    Proxy(WebCore::AudioMediaStreamTrackRendererInternalUnit::RenderCallback&&, WebCore::AudioMediaStreamTrackRendererInternalUnit::ResetCallback&&);
+    Proxy(Client&);
```
I suggest `explicit`.

darinadler commented on Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
```diff
@@ -59,8 +60,10 @@ class RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit
-    OSStatus render(size_t sampleCount, AudioBufferList&, uint64_t sampleTime, double hostTime, AudioUnitRenderActionFlags&);
-    void notifyReset();
+
+    // WebCore::AudioMediaStreamTrackRendererInternal::Client
```
Can these be private?

kkinnunen-apple replied:
These are public in the interface that defines them, `WebCore::AudioMediaStreamTrackRendererInternal::Client`.

I find it flawed practice to hide the overrides as private. This is on the grounds that the practice intends to communicate "these cannot be called by others", when in fact the sole point of the functions is that they should be called by others. Contrary to private inheritance, there is no language feature enforcing the intention of the function hiding. This means that to me it's more confusing than clarifying to redeclare overrides in different access specifier as their original definition. 

Taking the liberty to merge without the suggestion. If this suggestion is a consistent project-wide practice (written down somewhere in the future by me?), I'm happy to do another patch moving them.

darinadler replied:
Using `private` in this case means “called only polymorphically, not directly through a pointer specifically of this class’s type”. We can debate how valuable that is, but it’s not meaningless and to me it doesn’t come across as misleading about how public the function is in the base class. In fact whether the function being overridden in the base class might change over time and I would not want to update the visibility in all the derived classes. It’s a personal preference of mine, to clarify which functions are intended to be called in what fashion, not truly a project standard. I find it particularly helpful when a function is not helpful called non-polymorphically, like a function `isC` that returns `true` in a class named `C`. But also I think it helps make clear what a class is by cutting down on public members and it’s consistent with a general “everything as private as possible” pattern that can help make future refactoring simpler.

kkinnunen-apple replied:
Understood.
I moved the base class to private inheritance and moved the functions to private block.

kkinnunen-apple committed:
Creation of remote audio rendering internal unit is needlessly complex
https://bugs.webkit.org/show_bug.cgi?id=248168
rdar://problem/102577160

Reviewed by Darin Adler.

It takes two Functions that capture the same creator instance.
The creation is configured with a third Function.

Replace the two Functions with just a normal Client interface.

Replace the global factory Function with a normal function pointer.

Rename the creating functions to be just normal Class::create(...).

* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererInternalUnit::setCreateFunction):
(WebCore::AudioMediaStreamTrackRendererInternalUnit::create):
(WebCore::AudioMediaStreamTrackRendererInternalUnit::createLocalInternalUnit): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::LocalAudioMediaStreamTrackRendererInternalUnit): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::retrieveFormatDescription): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::setAudioOutputDevice): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::start): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::stop): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::createAudioUnitIfNeeded): Deleted.
(WebCore::clipAudioBuffer): Deleted.
(WebCore::clipAudioBufferList): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::render): Deleted.
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::renderingCallback): Deleted.
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererUnit::AudioMediaStreamTrackRendererUnit):
(WebCore::AudioMediaStreamTrackRendererUnit::render):
(WebCore::getCreateInternalUnitFunction): Deleted.
(WebCore::AudioMediaStreamTrackRendererUnit::setCreateInternalUnitFunction): Deleted.
(WebCore::createInternalUnit): Deleted.
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.h:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::Unit):
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::reset):
(WebKit::renderCallback): Deleted.
(WebKit::resetCallback): Deleted.
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::notifyReset): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::createRemoteAudioMediaStreamTrackRendererInternalUnitProxy):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::Proxy):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::startThread):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::reset):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::createRemoteInternalUnit): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.h:
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::setupCaptureProcesses):

Canonical link: https://commits.webkit.org/256969@main
